metabase/metabase
# add basic typescript configuration

### Description

This PR introduces Typescript configuration that will allow us to write new code in Typescript and gradually convert the existing code. It should not add significant overhead and at the same time, static type checks should prevent an entire class of bugs. Furthermore, in the long run, I believe it will speed up the development by creating type documentation for ourselves that will be used for autocomplete, code navigation, etc.

#### Key details:
- Babel is responsible for Typescript transpilation using `@babel/preset-typescript` preset. It does not check types.
- `yarn type-check` runs the Typescript compiler and checks actual types
- `yarn lint` now runs `yarn type-check` in addition to other checks

### How to verify
- Ensure the app works the same as it was before in dev/prod modes
- Ensure any type errors in `.ts` and `.tsx` files lead to errors listed by `yarn type-check`
- Check `tsconfig.json` and extended `eslint` configs make sense
